src/olympia/search/filters.py

```diff
@@ -105,6 +112,72 @@ def filter_queryset(self, request, qs, view):
             should=primary_should + secondary_should), functions=functions)
 
 
+class SearchParameterFilter(BaseFilterBackend):
+    """
+    A django-rest-framework filter backend that filters only items in an ES
+    query that match a specific set of fields: app, platform and type.
+    """
+    # FIXME: add appversion.
+
+    Filter = namedtuple(
+        'Filter', ['query_param', 'valid_values', 'reverse', 'es_field'])
+    available_filters = [
+        Filter(query_param='app', valid_values=amo.APP_IDS,
+               reverse=amo.APPS, es_field='app'),
+        Filter(query_param='platform', valid_values=amo.PLATFORMS,
+               reverse=amo.PLATFORM_DICT, es_field='platforms'),
+        Filter(query_param='type', valid_values=amo.ADDON_SEARCH_TYPES,
+               reverse=amo.ADDON_SEARCH_SLUGS, es_field='type'),
+    ]
+
+    def filter_queryset(self, request, qs, view):
+        must = []
+
+        for filter_instance in self.available_filters:
+            operator = 'term'
+            if filter_instance.query_param in request.GET:
+                value = request.GET[filter_instance.query_param]
+                try:
+                    # Try the int first.
+                    value = int(value)
+                except ValueError:
+                    # Fall back on the string, which needs to be in the reverse
+                    # dict.
+                    value = filter_instance.reverse.get(value.lower())
+                    # If our reverse dict contains objects and not integers,
+                    # like for platforms and apps, we need an extra step to
+                    # find the integer value.
+                    if hasattr(value, 'id'):
+                        value = value.id
+                if value in filter_instance.valid_values:
+                    # Small hack to support the fact that an add-on with
+                    # PLATFORM_ALL set supports any platform: when filtering
+                    # by any platform, always include PLATFORM_ALL. This means
+                    # we need to change the operator to use 'terms' instead of
+                    # 'term'.
+                    if (filter_instance.query_param == 'platform' and
+                            value != amo.PLATFORM_ALL.id):
+                        value = [value, amo.PLATFORM_ALL.id]
+                        operator = 'terms'
+                    must.append(
+                        F(operator, **{filter_instance.es_field: value}))
+
```

Comment:
Somehow I'm torn between liking and disliking this bit. There's too much guessing going on which could, or maybe even should be part of the specific `Filter` definition. That would add quite a bit of bloat though but maybe that's not a bad thing. Something like

```python
class AppFilter(object):
    query_param = 'app'
    reverse = amo.APPS
    valid_values = amo.APP_IDS
    es_field = 'app'

    def get_value(self, filter, value):
        value = self.reverse.get(value.lower())
        return value.id
```

This is just me quickly writing something down, but maybe this opens up more possibilities in the future for more complex lookups and avoids the actual filter guessing so much.

This could also inherit the `operator` value (the whole "hack" in https://github.com/mozilla/addons-server/pull/2548/files#diff-0fdef4f26957b30049ec712d6c470d15R152 could go away) and maybe implement something like `get_filter` but that's maybe too much.